# Исследование надежности заемщиков


**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
#print(data['education'].unique())
data['education'] = data['education'].str.lower()
print(data['education'].unique())

['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']


**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data['family_status'] = data['family_status'].str.lower() #для красоты и порядка

In [23]:
children_grouped = data['children'].value_counts()  # суммируем общее кол-во заемщиков для каждого числа детей
#print(children_grouped)
children_grouped_debt = data.groupby(data['children'])['debt'].value_counts() # суммируем число просрочунов в каждой группе
children_grouped_debt # нам нужно сделать функцию, которая будет считать эти значения, чтобы посчитать кол-во 
                      # просрочунов с определенным числом детей
aggfunc=['count', 'sum', 'mean'] # агрегация по подсказке 
children_table = data.groupby('children')['debt'].agg(aggfunc)

rang = children_table['sum'].count()
for index in range(rang): # В этом случае можно сразу вывести искомую вероятность на экран
    i = children_table.loc[index, 'mean']
    print(f'Для заемщиков с количеством детей, равным {index}, вероятность просроченной задолженности = {i:.2%} %') 
# ... и содержимое всех дальнейших ячеек в этом пункте в принципе можно удалить :)), и переходить к выводам


Для заемщиков с количеством детей, равным 0, вероятность просроченной задолженности = 7.54% %
Для заемщиков с количеством детей, равным 1, вероятность просроченной задолженности = 9.23% %
Для заемщиков с количеством детей, равным 2, вероятность просроченной задолженности = 9.45% %
Для заемщиков с количеством детей, равным 3, вероятность просроченной задолженности = 8.18% %
Для заемщиков с количеством детей, равным 4, вероятность просроченной задолженности = 9.76% %
Для заемщиков с количеством детей, равным 5, вероятность просроченной задолженности = 0.00% %


In [26]:
sum_debt_1 =[]  # объявляем пустой список сумм для кол-ва просрочунов с детьми
sorted_children = data['children'].unique()
print(sorted_children) # значения еще не отсортированы
sorted_children = sorted(sorted_children)
#print(sorted_children)

for i in sorted_children:# в списке заполняем список.
    sum = child_vs_debt(i, 1)
    sum_debt_1.append(sum) 
    print(f'Количество заемщиков с {i} детьми  равно {children_grouped[i]}')
    print(f'Количество просрочников с {i} детьми равно {sum}')    
    print('-------------------------------------------')

[1 0 3 2 4 5]
Количество заемщиков с 0 детьми  равно 14091
Количество просрочников с 0 детьми равно 1063
-------------------------------------------
Количество заемщиков с 1 детьми  равно 4808
Количество просрочников с 1 детьми равно 444
-------------------------------------------
Количество заемщиков с 2 детьми  равно 2052
Количество просрочников с 2 детьми равно 194
-------------------------------------------
Количество заемщиков с 3 детьми  равно 330
Количество просрочников с 3 детьми равно 27
-------------------------------------------
Количество заемщиков с 4 детьми  равно 41
Количество просрочников с 4 детьми равно 4
-------------------------------------------
Количество заемщиков с 5 детьми  равно 9
Количество просрочников с 5 детьми равно 0
-------------------------------------------


In [27]:
final_count_child = sum_debt_1 / children_grouped 
#final_count_child
index = 0
for i in final_count_child:
    print(f'Для заемщиков с количеством детей, равным {index}, вероятность просроченной задолженности = {i:.2%} %') 
    index +=1
    


Для заемщиков с количеством детей, равным 0, вероятность просроченной задолженности = 7.54% %
Для заемщиков с количеством детей, равным 1, вероятность просроченной задолженности = 9.23% %
Для заемщиков с количеством детей, равным 2, вероятность просроченной задолженности = 9.45% %
Для заемщиков с количеством детей, равным 3, вероятность просроченной задолженности = 8.18% %
Для заемщиков с количеством детей, равным 4, вероятность просроченной задолженности = 9.76% %
Для заемщиков с количеством детей, равным 5, вероятность просроченной задолженности = 0.00% %


**Вывод:** Зависимость между количеством детей и возвратом кредита в срок существует. Судя по результатам, самые надежные заемщики - многодетные с 5 детьми, но их мало (выборка нерепрезентативна), поэтому нецелесообразно принимать их в расчет. Кроме того, заемщиков с 4 детьми тоже не так много, лучше их не брать в расчет. Если их не учитывать, то самые надежные заемщики - не имеющие детей, с вероятностью дефолта 7,54%, в то время, как с увеличением количества детей растет и вероятность дефолта. Однако, среди имеющих детей самые надежные заемщики - многодетные с 3 детьми.

In [ ]:
fam_status_grouped = data.sort_values(by='family_status_id')
fam_status_grouped = fam_status_grouped['family_status_id'].value_counts()  # суммируем общее кол-во заемщиков для каждого числа детей
#print(fam_status_grouped)

fam_status_grouped_debt = data.groupby(data['family_status'])['debt'].value_counts() # суммируем число просрочунов в каждой группе
fam_status_grouped_debt # нам нужно сделать функцию, которая будет считать эти значения, чтобы посчитать кол-во 
                      # просрочунов в каждой группе 
#print('fam_status_grouped_debt=', fam_status_grouped_debt)


family_status          debt
в разводе              0        1105
                       1          84
вдовец / вдова         0         888
                       1          63
гражданский брак       0        3749
                       1         385
женат / замужем        0       11334
                       1         927
не женат / не замужем  0        2523
                       1         273
Name: debt, dtype: int64

In [ ]:
def fam_status_vs_debt(family_status_id, debt):
    # Объявляется функция с двумя параметрами: family_status, debt.
    # В переменной fam_status_debt_list сохраняются те строки таблицы data, для которых
    # значение в столбце 'family_status_id' равно параметру family_status и одновременно значение
    # в столбце 'debt' равно параметру debt (используем последовательную фильтрацию
    # с помощью логической индексации).
    fam_status_debt_list = data[data['family_status_id'] == family_status_id]
    #print(fam_status_debt_list['family_status'].count())
    fam_status_debt_list = fam_status_debt_list[fam_status_debt_list['debt'] == debt]
    #print(fam_status_debt_list['family_status_id'].count())

    # В переменной fam_status_is_debt_count сохраняется число значений столбца 'family_status_id',
    # рассчитанное методом count()
    fam_status_is_debt_count = fam_status_debt_list['family_status_id'].count()

    # Функция возвращает число - значение family_status_is_debt_count.
    return fam_status_is_debt_count

In [ ]:
fam_status_vs_debt(0,1) # проверяем, что все работает

927

In [ ]:
# делаем часть датасета справочник:
data_fam_dict = data[['family_status_id', 'family_status']]
data_fam_dict = data_fam_dict.drop_duplicates().reset_index(drop=True)

# выводим на экран кол-во заемщиков и должников по группам семейного положения
final_count_fam_status = [] #создаем список процентов к общему числу просрочунов в группах
for i in data_fam_dict['family_status_id']:  
    fam_sum = fam_status_vs_debt(i,1) #фильтруем только по должникам, у кого debt =1
    fs = data_fam_dict.loc[i,'family_status'] # текстовое значение группы семейного положения для последующего вывода в принте
    print('fam_status_grouped[i]', fam_status_grouped[i])
    final_count_fam_status.append(fam_sum / fam_status_grouped[i]) # считаем процент должников от общего количества
    print(f'Для заемщиков группы {i} - "{fs}", вероятность просроченной задолженности = {final_count_fam_status[i]:.2%} %') 
    #print('-------------------------------------------')
    

fam_status_grouped[i] 12261
Для заемщиков группы 0 - "женат / замужем", вероятность просроченной задолженности = 7.56% %
fam_status_grouped[i] 4134
Для заемщиков группы 1 - "гражданский брак", вероятность просроченной задолженности = 9.31% %
fam_status_grouped[i] 951
Для заемщиков группы 2 - "вдовец / вдова", вероятность просроченной задолженности = 6.62% %
fam_status_grouped[i] 1189
Для заемщиков группы 3 - "в разводе", вероятность просроченной задолженности = 7.06% %
fam_status_grouped[i] 2796
Для заемщиков группы 4 - "не женат / не замужем", вероятность просроченной задолженности = 9.76% %


**Вывод:** Самые надежные заемщики - из группы "вдовец/вдова" с вероятностью выхода на просрочку 6.62%, а также заемщики из группы "в разводе", для которых эта вероятность равна 7,06%. Самые недобросовестные заемщики - неженатые/незамужние (9,76%).

In [ ]:
income_grouped = data['total_income_category'].value_counts()  # суммируем общее кол-во заемщиков для каждой группы пол доходам
print(income_grouped)
income_grouped_debt = data.groupby(data['total_income_category'])['debt'].value_counts() # суммируем число просрочунов в каждой группе
income_grouped_debt# нам нужно сделать функцию, которая будет считать эти значения, чтобы посчитать кол-во 
                      # просрочунов в каждой группе

C    15921
B     5014
D      349
A       25
E       22
Name: total_income_category, dtype: int64


total_income_category  debt
A                      0          23
                       1           2
B                      0        4660
                       1         354
C                      0       14568
                       1        1353
D                      0         328
                       1          21
E                      0          20
                       1           2
Name: debt, dtype: int64

In [ ]:
def income_category_vs_debt(total_income_category, debt):
    # Объявляется функция с двумя параметрами: total_income_category, debt.
    # В переменной income_category_debt_list сохраняются те строки таблицы data, для которых
    # значение в столбце 'total_income_category' равно параметру total_income_category и одновременно значение
    # в столбце 'debt' равно параметру debt (используем последовательную фильтрацию
    # с помощью логической индексации).
    income_category_debt_list = data[data['total_income_category'] == total_income_category]
    income_category_debt_list = income_category_debt_list[income_category_debt_list['debt'] == debt]
    #print(income_category_debt_list['total_income_category'].count())

    # В переменной income_category_is_debt_count сохраняется число значений столбца 'total_income_category',
    # рассчитанное методом count()
    income_category_is_debt_count = income_category_debt_list['total_income_category'].count()

    # Функция возвращает число - значение income_category_is_debt_count.
    return income_category_is_debt_count

In [ ]:
income_category_vs_debt('C',1) # проверяем, что все работает

1353

In [ ]:
# делаем часть датасета- справочник:
income_text = ['свыше 1000 тыс. руб','от 200 до 1000 тыс. руб', 'от 50 до 200 тыс. руб', \
               'от 30 до 50 тыс. руб', 'до 30 тыс. руб']
data_income_dict = data['total_income_category']
data_income_dict = data_income_dict.drop_duplicates().reset_index(drop=True)
data_income_dict = data_income_dict.sort_values()
data_income_dict = pd.DataFrame(list(zip(data_income_dict, income_text)),
               columns =['total_income_category', 'description'])
print(data_income_dict)

  total_income_category              description
0                     A      свыше 1000 тыс. руб
1                     B  от 200 до 1000 тыс. руб
2                     C    от 50 до 200 тыс. руб
3                     D     от 30 до 50 тыс. руб
4                     E           до 30 тыс. руб


In [ ]:
# выводим на экран кол-во заемщиков и должников по группам дохода
final_count_income_category = [] #создаем список процентов к общему числу просрочунов в группах
index = 0
for i in data_income_dict['total_income_category']:  
    #print('i=', i)
    income_sum = income_category_vs_debt(i,1) #фильтруем только по должникам, у кого debt =1
    #print('income_sum=',income_sum)
    description = data_income_dict.loc[index,'description'] # текстовое значение группы дохода для последующего вывода в принте
    #print('index=', index)
    #print('description=', description)
    inc = income_grouped.loc[i] # общее количество значений в группе i
    #print('income_grouped[index]', inc)
    final_count_income_category.append(income_sum / inc) # считаем процент должников от общего количества
    print(f'Для заемщиков группы {i} - "{description}", вероятность просроченной задолженности \
    = {final_count_income_category[index]:.2%} %') 
    index +=1

Для заемщиков группы A - "свыше 1000 тыс. руб", вероятность просроченной задолженности     = 8.00% %
Для заемщиков группы B - "от 200 до 1000 тыс. руб", вероятность просроченной задолженности     = 7.06% %
Для заемщиков группы C - "от 50 до 200 тыс. руб", вероятность просроченной задолженности     = 8.50% %
Для заемщиков группы D - "от 30 до 50 тыс. руб", вероятность просроченной задолженности     = 6.02% %
Для заемщиков группы E - "до 30 тыс. руб", вероятность просроченной задолженности     = 9.09% %


**Вывод:** С точки зрения доходов самые надежные заемщики - из группы D (30-50 тыс. руб), самые ненадежные - из группы Е (до 30 тыс. руб). 

In [ ]:
purpose_grouped = data['purpose_category'].value_counts()  # суммируем общее кол-во заемщиков для каждой группы пол доходам
print(purpose_grouped)
purpose_grouped_debt = data.groupby(data['purpose_category'])['debt'].value_counts() # суммируем число просрочунов в каждой группе
purpose_grouped_debt# нам нужно сделать функцию, которая будет считать эти значения, чтобы посчитать кол-во 
                      # просрочунов в каждой группе

операции с недвижимостью    10751
операции с автомобилем       4279
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64


purpose_category          debt
операции с автомобилем    0       3879
                          1        400
операции с недвижимостью  0       9971
                          1        780
получение образования     0       3619
                          1        369
проведение свадьбы        0       2130
                          1        183
Name: debt, dtype: int64

In [ ]:
def purpose_category_vs_debt(purpose_category, debt):
    # Объявляется функция с двумя параметрами: purpose_category, debt.
    # В переменной purpose_category_debt_list сохраняются те строки таблицы data, для которых
    # значение в столбце 'purpose_category' равно параметру purpose_category и одновременно значение
    # в столбце 'debt' равно параметру debt (используем последовательную фильтрацию
    # с помощью логической индексации).
    purpose_category_debt_list = data[data['purpose_category'] == purpose_category]
    purpose_category_debt_list = purpose_category_debt_list[purpose_category_debt_list['debt'] == debt]
    #print(purpose_category_debt_list['purpose_category'].count())

    # В переменной purpose_category_is_debt_count сохраняется число значений столбца 'purpose_category',
    # рассчитанное методом count()
    purpose_category_is_debt_count = purpose_category_debt_list['purpose_category'].count()

    # Функция возвращает число - значение income_category_is_debt_count.
    return purpose_category_is_debt_count

In [ ]:
data.purpose_category.unique()
#purpose_category_vs_debt('операции с автомобилем',1)

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

In [ ]:
# делаем часть датасета- справочник:
purpose_text = data.purpose_category.unique()
purpose_text

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

In [ ]:
# выводим на экран кол-во заемщиков и должников по группам целей кредитования
final_count_purpose_category = [] #создаем список процентов к общему числу просрочунов в группах
index = 0
for i in purpose_text:  
    print('i=', i)
    purpose_sum = purpose_category_vs_debt(i,1) #фильтруем только по должникам, у кого debt =1
    print('income_sum=',purpose_sum)
    purp = purpose_grouped.loc[i] # общее количество значений в группе i
    print('purp =', purp)
    final_count_purpose_category.append(purpose_sum / purp) # считаем процент должников от общего количества
    print(f'Для заемщиков группы "{i}" , вероятность просроченной задолженности \
    = {final_count_purpose_category[index]:.2%} %') 
    index +=1

i= операции с недвижимостью
income_sum= 780
purp = 10751
Для заемщиков группы "операции с недвижимостью" , вероятность просроченной задолженности     = 7.26% %
i= операции с автомобилем
income_sum= 400
purp = 4279
Для заемщиков группы "операции с автомобилем" , вероятность просроченной задолженности     = 9.35% %
i= получение образования
income_sum= 369
purp = 3988
Для заемщиков группы "получение образования" , вероятность просроченной задолженности     = 9.25% %
i= проведение свадьбы
income_sum= 183
purp = 2313
Для заемщиков группы "проведение свадьбы" , вероятность просроченной задолженности     = 7.91% %


**Вывод:** С точки зрения целей кредитования - самые надежные заемщики берут кредит на операции с недвижимостью (вероятность дефолта 7.26%), а также на проведение свадьбы (вероятность дефолта 7.91%). Ко второй группе с примерно одинаковой повышенной вероятностью дефолта (9.35% и 9.25%) относятся кредиты на операции с авто и на получение образования.

В данном датасете изначально встречались пропуски в столбцах days_employed и total_income. Возникновение пропусков в  столбце total_income, вероятно, связано с нежеланием заемщика указывать конкретную сумму своего дохода, либо ошибками при выгрузке/сохранении данных; в столбце days_employed, который, скорее всего заполняется не клиентом, а автоматически - техническими ошибками при сохранеии/выгрузке данных.  

Заполнение пропусков для количественных переменных медианой является лучшим решением из-за устойчивости медианы к выбросам и аномальным значениям, которые, как правило, присутствуют в выборке количественных переменных.

### Шаг 4: общий вывод.

В качестве заказчика исследования выступил кредитный отдел банка. Целью исследования является установление зависимостей между определенными характеристиками заемщиков (например, кол-во детей, семейное положение и проч.) и кредитной историей (положительная/отрицательная). Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования планируется использовать при построении модели кредитного скоринга для потенциальных клиентов.

Исходные данные, содержали были неидеальны, поэтому потребовалась их предобработка:
1. Данные содержали пропуски в двух столбцах: "кол-во отработанных дней",а также "уровнень дохода". В каждом из этих столбцов пропуски были заменены медианными значениями  (столбец "кол-во отработанных дней" - по каждому типу занятости).
2. В столбце "кол-во отработанных дней" были отрицательные значения (были заменены на положительные).
3. В столбце "количество детей" были аномальные значения - отрицательное число, и слишком большое(20) - оба удалены (датасет отфильтрован).
4. Столбец "образование" содержал неявные дубликаты, которые пришлось ликвидировать приведением значений к нижнему регистру.
5. Кроме того, датасет содержал строки - полные дубликаты (71 шт) - удалены.
6. Данные были категоризированы по уровню дохода (в 5 категорий), а также по целям кредитования (в 4 категории).

На этапе анализа были проверены гипотезы:
- о наличии зависимости между количеством детей и возвратом кредита в срок,
- о наличии зависимости между семейным положением и возвратом кредита в срок,
- о наличии зависимости между уровнем дохода и возвратом кредита в срок,
- о том, что разные цели кредита также влияют на уровень просроченной задолженности.


В результате, сделаны выводы:

1. Зависимость между количеством детей и возвратом кредита в срок существует. Судя по результатам, самые надежные заемщики - многодетные с 5 детьми, но их мало (выборка нерепрезентативна), поэтому нецелесообразно принимать их в расчет. Если их не учитывать, то самые надежные заемщики - не имеющие детей, с вероятностью дефолта 7,54%, в то время, как с увеличением количества детей растет и вероятность дефолта. Однако, среди имеющих детей самые надежные заемщики - многодетные с 3 детьми.

2. Самые надежные заемщики с точки зрения семейного положения - из группы "вдовец/вдова" с вероятностью выхода на просрочку 6.62%, а также заемщики из группы "в разводе", для которых эта вероятность равна 7,06%. Самые недобросовестные заемщики - неженатые/незамужние (9,76%).

3. С точки зрения доходов самые надежные заемщики - из группы D (30-50 тыс. руб), самые ненадежные - из группы Е (до 30 тыс. руб).
 
4. С точки зрения целей кредитования - самые надежные заемщики берут кредит на операции с недвижимостью (вероятность дефолта 7.26%), а также на проведение свадьбы (вероятность дефолта 7.91%). Ко второй группе с примерно одинаковой повышенной вероятностью дефолта (9.35% и 9.25%) относятся кредиты на операции с авто и на получение образования.

В результате исследования можно сформировать портрет "идеального" заемщика:

Идеальный заемщик - это бездетный(ая) вдовец/вдова с уровнем дохода от 30 до 50 тыс. руб., берущий кредит на операции с недвижимостью или проведение свадьбы.


Рекомендации по улучшению сбора данных:
1. Сделать поле ввода уровня дохода обязательным при заполнении анкеты.
2. Корректно выгружать / сохранять данные о трудовом стаже.
3. В таких графах как "количество детей", "цели кредитования" давать заемщикам выбирать из нескольких вариантов, а не вносить произвольное значение.


